
this reverts:

From c6bba31dbdd450081503dc0e20fcfb4328b347e4 Mon Sep 17 00:00:00 2001
From: Jonas Karlman <jonas@kwiboo.se>
Date: Sat, 12 Jul 2025 21:12:29 +0000
Subject: [PATCH] rockchip: spl-boot-order: Defer probe of boot device

Boot devices are being probed when SPL boot order is determined. This
may delay boot slightly and can prevent booting from SPI Flash on boards
that use same pins for SPI Flash and eMMC due to pinctrl being applied
prior to booting.

Instead defer probe of the boot device until SPL try to load image from
the boot device by using uclass_find_device_by_of_offset() instead of
the get variant.

wihtout this revert, boot of e.g rock5-itx oops like this:

[   25.198501] Run /sbin/init as init process
[   25.199187]   with arguments:
[   25.200028]     /sbin/init
[   25.200528]   with environment:
[   25.201025]     HOME=/
[   25.201457]     TERM=linux
[   25.691079] rockchip-drm display-subsystem: bound fdd90000.vop (ops vop2_component_ops)
[   25.709451] Unable to handle kernel NULL pointer dereference at virtual address 0000000000000100
[   25.710253] Mem abort info:
[   25.710509]   ESR = 0x0000000096000006
[   25.710849]   EC = 0x25: DABT (current EL), IL = 32 bits
[   25.711328]   SET = 0, FnV = 0
[   25.711607]   EA = 0, S1PTW = 0
[   25.711894]   FSC = 0x06: level 2 translation fault
[   25.712332] Data abort info:
[   25.712595]   ISV = 0, ISS = 0x00000006, ISS2 = 0x00000000
[   25.713085]   CM = 0, WnR = 0, TnD = 0, TagAccess = 0
[   25.713540]   GCS = 0, Overlay = 0, DirtyBit = 0, Xs = 0
[   25.714018] user pgtable: 4k pages, 48-bit VAs, pgdp=000000010181c000
[   25.714594] [0000000000000100] pgd=0800000101824403, p4d=0800000101824403, pud=0800000101825403, pmd=0000000000000000
[   25.715564] Internal error: Oops: 0000000096000006 [#1]  SMP
[   25.716074] Modules linked in:
[   25.716364] CPU: 0 UID: 0 PID: 113 Comm: kworker/u32:7 Not tainted 6.18.5 #1 PREEMPT  4b1cc0630507d829b801454d4615642bb605b542
[   25.717379] Hardware name: Radxa ROCK 5 ITX (DT)
[   25.717802] Workqueue: events_unbound deferred_probe_work_func
[   25.718350] pstate: 804000c9 (Nzcv daIF +PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[   25.718977] pc : get_partial_node+0x6c/0x1e8
[   25.719374] lr : get_partial_node+0x3c/0x1e8
[   25.719767] sp : ffff800082ef3880
[   25.720068] x29: ffff800082ef3880 x28: ffff0003fc524000 x27: ffff0003fc524020
[   25.720727] x26: 0000000000000001 x25: ffff000100818010 x24: 00000000ffffffff
[   25.721385] x23: 0000000000000130 x22: ffff800082ef38e8 x21: ffff0003fc501780
[   25.722043] x20: ffff0003fc500280 x19: fffffdffc40aac00 x18: 00000005efd474ca
[   25.722701] x17: 000000040044ffff x16: 001000f2b5503510 x15: 0000000000000000
[   25.723357] x14: ffff8000818f3200 x13: ffff0001027b00d0 x12: 0000000000000001
[   25.724014] x11: 00000000000000c0 x10: 000000000008e410 x9 : 0000000000000004
[   25.724670] x8 : 0101010101010101 x7 : 00000000736c6c65 x6 : 0000000000000000
[   25.725326] x5 : fffffdffc4011000 x4 : 0000000000000130 x3 : 0000000000000100
[   25.725983] x2 : ffff800082ef38e8 x1 : 0000000000000000 x0 : 0000000000000000
[   25.726641] Call trace:
[   25.726868]  get_partial_node+0x6c/0x1e8 (P)
[   25.727267]  __slab_alloc_node.isra.0+0x14c/0x2c8
[   25.727701]  slab_alloc_node.isra.0+0x144/0x180
[   25.728120]  __kmalloc_cache_noprof+0x1c/0x24
[   25.728524]  dev_pm_qos_constraints_allocate+0x2c/0xe4
[   25.729001]  dev_pm_qos_add_notifier+0x130/0x160
[   25.729432]  genpd_add_device+0x16c/0x2e0
[   25.729811]  __genpd_dev_pm_attach+0xac/0x308
[   25.730219]  genpd_dev_pm_attach+0x60/0x78
[   25.730604]  dev_pm_domain_attach+0x68/0x80
[   25.730995]  platform_probe+0x44/0xa0
[   25.731338]  really_probe+0xbc/0x384
[   25.731677]  __driver_probe_device+0x78/0x148
[   25.732085]  driver_probe_device+0x3c/0x118
[   25.732476]  __device_attach_driver+0xb0/0x158
[   25.732891]  bus_for_each_drv+0x80/0xdc
[   25.733252]  __device_attach+0x98/0x1b4
[   25.733613]  device_initial_probe+0x10/0x18
[   25.734006]  bus_probe_device+0x94/0x98
[   25.734367]  deferred_probe_work_func+0xb4/0x120
[   25.734796]  process_one_work+0x164/0x2ac
[   25.735171]  worker_thread+0x2bc/0x3e4
[   25.735521]  kthread+0x100/0x1b8
[   25.735826]  ret_from_fork+0x10/0x20
[   25.736175] Code: f9400263 374004c3 f9401266 b9402aa3 (f86368c3) 
[   25.736721] ---[ end trace 0000000000000000 ]---

diff --speed-large-files --no-dereference --minimal -Naur u-boot-127a42c7257a6ffbbd1575ed1cbaa8f5408a44b3/arch/arm/mach-rockchip/spl-boot-order.c u-boot-127a42c7257a6ffbbd1575ed1cbaa8f5408a44b3/arch/arm/mach-rockchip/spl-boot-order.c
--- u-boot-127a42c7257a6ffbbd1575ed1cbaa8f5408a44b3/arch/arm/mach-rockchip/spl-boot-order.c	2026-01-05 21:49:22.000000000 +0100
+++ u-boot-127a42c7257a6ffbbd1575ed1cbaa8f5408a44b3/arch/arm/mach-rockchip/spl-boot-order.c	2026-01-22 18:58:53.386725299 +0100
@@ -41,7 +41,7 @@
 	 * aware of the block-device layer.  Until then (and to avoid unneeded
 	 * delays in getting this feature out), it lives at the board-level.
 	 */
-	if (!uclass_find_device_by_of_offset(UCLASS_MMC, node, &parent)) {
+	if (!uclass_get_device_by_of_offset(UCLASS_MMC, node, &parent)) {
 		struct udevice *dev;
 		struct blk_desc *desc = NULL;
 
@@ -73,7 +73,7 @@
 	 * extended with awareness of the BLK layer (and matching OF_CONTROL)
 	 * soon.
 	 */
-	if (!uclass_find_device_by_of_offset(UCLASS_SPI_FLASH, node, &parent))
+	if (!uclass_get_device_by_of_offset(UCLASS_SPI_FLASH, node, &parent))
 		return BOOT_DEVICE_SPI;
 
 	return -1;
